---
标题: 5.defer、recover异常捕获和处理
创建时间: 2023-03-28 17:37
修改时间: 2023-03-29 10:59
tags: 
---


# 🥏defer的应用场景
## 1. 资源释放
> 关闭文件、关闭连接等
## 2. 异常捕获和处理

# 🕤defer
>1. defer关键字用来声明一个延迟调用的函数，该函数可以使匿名函数也可以是具名函数。
>2. defer延迟函数执行时机，方法return之后，返回参数到调用方法之前。
>3. defer延迟函数可以在方法返回值后改变函数的返回值。
>4. 在方法结束（正常返回，异常结束）后，都会调用defer声明的延迟函数，可以有效避免因异常导致的资源无法释放的问题。
>5. 可以指定多个defer延迟函数，多个函数执行顺序为后进先出。
>6. defer与recover配合可以实现异常捕获与处理逻辑。
>7. 不建议在for循环中使用defer

# 👓recover
>1. Go语言的内建函数，可以让进入宕机流程中的goroutine恢复过来。
>2. recover仅在延迟函数defer中有效，在正常执行过程中，调用recover会返回nil并没有其他任何效果。
>3. 如果当前的goroutine出现panic，调用recover可以捕获到panic的输入值，并恢复正常的执行。

# 🚨panic
>1. Go语言的一种异常机制。
>2. 可以通过panic函数主动抛出异常

```go
func f1() {
	println("in defer function2")
}

func deferCase1() {
	println("start defer case 1")

	defer func() {
		println("in defer function1")
	}()
	defer f1()
	println("defer case end")
}
```
执行结果如下:
```sh
start defer case 1
defer case end
in defer function2
in defer function1
```

```go
//参数预计算
func deferCase2(){
  i := 1 
  //传参
  defer func(j int){
    println("defer j:",j)
  }(i+1)
  //闭包
  defer func(){
    println("defer j:",i)
  }()
  i = 99
  println("i: ",i)
}
```
运行结果如下:
```sh
i:  99
defer j: 99
defer j: 2
```

```go
//返回值，defer执行在returun之后
var g = 100
func f2()(int,*int){
  defer func(){
    g = 200
  }()
  println("f2 g: ",g)
  return g,&g
}

func deferCase3(){
  i,j := f2()
  println("i,j,g: ",i,*j,g)
}
```
运行结果如下:
```sh
f2 g:  100
i,j,g:  100 200 200
```

使用defer和recover进行异常处理
```go
func exceptionCase(){
  defer func(){
    err := recover()
    if err != nil{
      println("error dealing...  defer recover:",err)
    }
  }()
  println("start exceptionCase")
  panic("exceptionCase error")
  println("end exceptionCase")
}
```
运行结果:
```sh
start exceptionCase
error dealing...  defer recover: (0x45d6a0,0x47c610)
```
defer用于释放资源
```go
func resorceCloseCase(){
  file,err := os.Open("READ.md")
  if err != nil {
    log.Fatal(err)
  }
  defer file.Close()
}
```
> 注意，这里defer最好是放在err的后面，以防止file为nil时调用Close。